File: scripts/promote_submission.py
#!/usr/bin/env python3
"""
Publish a fragrance submission into the shared catalog (in place).

Usage:
  python scripts/promote_submission.py --id 123

Prefers the provisional fragrance linked via promoted_fragrance_id (migration 005).
Publishes that same row with a MiniLM embedding — collection FKs stay unchanged.
Falls back to insert if no linked provisional exists (legacy submissions).
"""
from __future__ import annotations

import argparse
import os
from datetime import datetime, timezone
from pathlib import Path

from dotenv import load_dotenv
from huggingface_hub import login
from sentence_transformers import SentenceTransformer
from supabase import create_client

load_dotenv(Path(__file__).resolve().parent / ".env")
if token := os.environ.get("HF_TOKEN"):
    login(token)

MODEL = SentenceTransformer("all-MiniLM-L6-v2")


def build_embedding_text(row: dict) -> str:
    parts = [
        row.get("main_accord_1"),
        row.get("main_accord_2"),
        row.get("main_accord_3"),
        row.get("main_accord_4"),
        row.get("main_accord_5"),
        row.get("top_notes"),
        row.get("middle_notes"),
        row.get("base_notes"),
    ]
    joined = " ".join(str(p).strip() for p in parts if p and str(p).strip())
    if joined:
        return joined
    return f"{row.get('perfume', '')} {row.get('brand', '')}".strip()


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Publish a submission to the shared catalog (in place)"
    )
    parser.add_argument("--id", type=int, required=True, help="fragrance_submissions.id")
    parser.add_argument(
        "--rating",
        type=float,
        default=4.0,
        help="rating_value when inserting a legacy row (ignored if publishing provisional)",
    )
    parser.add_argument("--reviewed-by", default="admin", help="Reviewer label")
    args = parser.parse_args()

    url = os.environ.get("SUPABASE_URL")
    key = os.environ.get("SUPABASE_SERVICE_ROLE_KEY")
    if not url or not key:
        raise SystemExit("Missing SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY in scripts/.env")

    sb = create_client(url, key)

    submission = (
        sb.table("fragrance_submissions")
        .select("*")
        .eq("id", args.id)
        .single()
        .execute()
        .data
    )
    if not submission:
        raise SystemExit(f"Submission {args.id} not found")

    if submission["status"] == "approved":
        raise SystemExit(f"Submission {args.id} is already approved")

    fragrance = None
    promoted_id = submission.get("promoted_fragrance_id")
    if promoted_id:
        fragrance = (
            sb.table("fragrances")
            .select("*")
            .eq("id", promoted_id)
            .single()
            .execute()
            .data
        )

    if not fragrance:
        existing = (
            sb.table("fragrances")
            .select("*")
            .eq("perfume", submission["perfume"])
            .eq("brand", submission["brand"])
            .execute()
            .data
        )
        fragrance = existing[0] if existing else None

    embed_src = fragrance or submission
    embed_text = build_embedding_text(embed_src)
    print(f"Embedding: {embed_text[:120]}…")
    embedding = MODEL.encode([embed_text])[0].tolist()
    reviewed_at = datetime.now(timezone.utc).isoformat()

    if fragrance:
        fragrance_id = fragrance["id"]
        sb.table("fragrances").update(
            {
                "visibility": "published",
                "embedding": embedding,
                "country": fragrance.get("country") or submission.get("country"),
                "gender": fragrance.get("gender") or submission.get("gender"),
                "top_notes": fragrance.get("top_notes") or submission.get("top_notes"),
                "middle_notes": fragrance.get("middle_notes")
                or submission.get("middle_notes"),
                "base_notes": fragrance.get("base_notes") or submission.get("base_notes"),
                "main_accord_1": fragrance.get("main_accord_1")
                or submission.get("main_accord_1"),
                "main_accord_2": fragrance.get("main_accord_2")
                or submission.get("main_accord_2"),
                "main_accord_3": fragrance.get("main_accord_3")
                or submission.get("main_accord_3"),
                "main_accord_4": fragrance.get("main_accord_4")
                or submission.get("main_accord_4"),
                "main_accord_5": fragrance.get("main_accord_5")
                or submission.get("main_accord_5"),
            }
        ).eq("id", fragrance_id).execute()
        print(f"Published fragrance id={fragrance_id} (in place)")
    else:
        row = {
            "perfume": submission["perfume"],
            "brand": submission["brand"],
            "country": submission.get("country"),
            "gender": submission.get("gender"),
            "rating_value": args.rating,
            "rating_count": 0,
            "top_notes": submission.get("top_notes"),
            "middle_notes": submission.get("middle_notes"),
            "base_notes": submission.get("base_notes"),
            "main_accord_1": submission.get("main_accord_1"),
            "main_accord_2": submission.get("main_accord_2"),
            "main_accord_3": submission.get("main_accord_3"),
            "main_accord_4": submission.get("main_accord_4"),
            "main_accord_5": submission.get("main_accord_5"),
            "embedding": embedding,
            "visibility": "published",
        }
        inserted = sb.table("fragrances").insert(row).execute().data
        if not inserted:
            raise SystemExit("Failed to insert into fragrances")
        fragrance_id = inserted[0]["id"]
        print(f"Inserted fragrance id={fragrance_id} with embedding")

    sb.table("fragrance_submissions").update(
        {
            "status": "approved",
            "reviewed_at": reviewed_at,
            "reviewed_by": args.reviewed_by,
            "promoted_fragrance_id": fragrance_id,
        }
    ).eq("id", args.id).execute()

    print(f"Submission {args.id} marked approved")


if __name__ == "__main__":
    main()

File: apps/web/lib/server/promote-submission.ts
/**
 * Publish a provisional submission in place (same fragrance_id).
 * Reject leaves the Custom provisional row in the user's collection.
 */

import { embedSearchQuery } from "@/lib/server/embed-query";
import {
  fragranceSelect,
  mapFragrance,
  type FragranceRow,
} from "@/lib/server/fragrance-mapper";
import {
  mapSubmission,
  submissionSelect,
  type SubmissionRow,
} from "@/lib/server/submission-mapper";
import { supabaseAdmin } from "@/lib/server/supabase-admin";
import type { Fragrance } from "@/lib/types/fragrance";
import type { FragranceSubmission } from "@/lib/types/submission";

export type PromoteResult =
  | {
      ok: true;
      submission: FragranceSubmission;
      fragrance: Fragrance;
    }
  | { ok: false; status: number; message: string };

function buildEmbeddingText(row: {
  perfume: string;
  brand: string;
  main_accord_1?: string | null;
  main_accord_2?: string | null;
  main_accord_3?: string | null;
  main_accord_4?: string | null;
  main_accord_5?: string | null;
  top_notes?: string | null;
  middle_notes?: string | null;
  base_notes?: string | null;
}): string {
  const parts = [
    row.main_accord_1,
    row.main_accord_2,
    row.main_accord_3,
    row.main_accord_4,
    row.main_accord_5,
    row.top_notes,
    row.middle_notes,
    row.base_notes,
  ];
  const joined = parts
    .map((p) => (p ? String(p).trim() : ""))
    .filter(Boolean)
    .join(" ");
  return joined || `${row.perfume} ${row.brand}`;
}

async function loadSubmission(id: number): Promise<SubmissionRow | null> {
  const { data, error } = await supabaseAdmin
    .from("fragrance_submissions")
    .select(submissionSelect)
    .eq("id", id)
    .maybeSingle();

  if (error) throw new Error(error.message);
  return (data as SubmissionRow | null) ?? null;
}

async function resolveFragranceForSubmission(
  submission: SubmissionRow,
): Promise<FragranceRow | null> {
  if (submission.promoted_fragrance_id != null) {
    const { data, error } = await supabaseAdmin
      .from("fragrances")
      .select(fragranceSelect)
      .eq("id", submission.promoted_fragrance_id)
      .maybeSingle();
    if (error) throw new Error(error.message);
    if (data) return data as FragranceRow;
  }

  // Legacy submissions (pre-005): find/create provisional by perfume+brand.
  const { data: existing, error: findError } = await supabaseAdmin
    .from("fragrances")
    .select(fragranceSelect)
    .eq("perfume", submission.perfume)
    .eq("brand", submission.brand)
    .maybeSingle();
  if (findError) throw new Error(findError.message);
  if (existing) return existing as FragranceRow;

  const { data: inserted, error: insertError } = await supabaseAdmin
    .from("fragrances")
    .insert({
      perfume: submission.perfume,
      brand: submission.brand,
      country: submission.country,
      gender: submission.gender,
      rating_value: 4.0,
      rating_count: 0,
      top_notes: submission.top_notes,
      middle_notes: submission.middle_notes,
      base_notes: submission.base_notes,
      main_accord_1: submission.main_accord_1,
      main_accord_2: submission.main_accord_2,
      main_accord_3: submission.main_accord_3,
      main_accord_4: submission.main_accord_4,
      main_accord_5: submission.main_accord_5,
      visibility: "provisional",
    })
    .select(fragranceSelect)
    .single();

  if (insertError || !inserted) {
    throw new Error(insertError?.message ?? "Failed to create fragrance row");
  }
  return inserted as FragranceRow;
}

export async function approveSubmission(
  submissionId: number,
  reviewedBy: string,
): Promise<PromoteResult> {
  const submission = await loadSubmission(submissionId);
  if (!submission) {
    return { ok: false, status: 404, message: "Submission not found." };
  }
  if (submission.status === "approved") {
    return { ok: false, status: 409, message: "Submission already approved." };
  }

  const fragrance = await resolveFragranceForSubmission(submission);
  if (!fragrance) {
    return {
      ok: false,
      status: 500,
      message: "Could not resolve linked fragrance.",
    };
  }

  const embedText = buildEmbeddingText({
    perfume: fragrance.perfume,
    brand: fragrance.brand,
    main_accord_1: fragrance.main_accord_1,
    main_accord_2: fragrance.main_accord_2,
    main_accord_3: fragrance.main_accord_3,
    main_accord_4: fragrance.main_accord_4,
    main_accord_5: fragrance.main_accord_5,
    top_notes: fragrance.top_notes,
    middle_notes: fragrance.middle_notes,
    base_notes: fragrance.base_notes,
  });

  const embedding = await embedSearchQuery(embedText);
  if (!embedding) {
    return {
      ok: false,
      status: 503,
      message:
        "Could not generate embedding (check HF_TOKEN). Approval aborted.",
    };
  }

  const { data: published, error: publishError } = await supabaseAdmin
    .from("fragrances")
    .update({
      visibility: "published",
      embedding,
      // Prefer submission metadata if fragrance was a thin provisional.
      country: fragrance.country ?? submission.country,
      gender: fragrance.gender ?? submission.gender,
      top_notes: fragrance.top_notes ?? submission.top_notes,
      middle_notes: fragrance.middle_notes ?? submission.middle_notes,
      base_notes: fragrance.base_notes ?? submission.base_notes,
      main_accord_1: fragrance.main_accord_1 ?? submission.main_accord_1,
      main_accord_2: fragrance.main_accord_2 ?? submission.main_accord_2,
      main_accord_3: fragrance.main_accord_3 ?? submission.main_accord_3,
      main_accord_4: fragrance.main_accord_4 ?? submission.main_accord_4,
      main_accord_5: fragrance.main_accord_5 ?? submission.main_accord_5,
    })
    .eq("id", fragrance.id)
    .select(fragranceSelect)
    .single();

  if (publishError || !published) {
    return {
      ok: false,
      status: 500,
      message: publishError?.message ?? "Failed to publish fragrance.",
    };
  }

  const { data: updatedSubmission, error: updateError } = await supabaseAdmin
    .from("fragrance_submissions")
    .update({
      status: "approved",
      reviewed_at: new Date().toISOString(),
      reviewed_by: reviewedBy,
      promoted_fragrance_id: Number(fragrance.id),
    })
    .eq("id", submissionId)
    .select(submissionSelect)
    .single();

  if (updateError || !updatedSubmission) {
    return {
      ok: false,
      status: 500,
      message: updateError?.message ?? "Failed to mark submission approved.",
    };
  }

  return {
    ok: true,
    submission: mapSubmission(updatedSubmission as SubmissionRow),
    fragrance: mapFragrance(published as FragranceRow),
  };
}

export async function rejectSubmission(
  submissionId: number,
  reviewedBy: string,
): Promise<PromoteResult> {
  const submission = await loadSubmission(submissionId);
  if (!submission) {
    return { ok: false, status: 404, message: "Submission not found." };
  }
  if (submission.status === "approved") {
    return {
      ok: false,
      status: 409,
      message: "Approved submissions cannot be rejected.",
    };
  }
  if (submission.status === "rejected") {
    return { ok: false, status: 409, message: "Submission already rejected." };
  }

  // Ensure a provisional fragrance exists for the submitter's collection link.
  const fragrance = await resolveFragranceForSubmission(submission);
  if (!fragrance) {
    return {
      ok: false,
      status: 500,
      message: "Could not resolve linked fragrance.",
    };
  }

  if ((fragrance.visibility ?? "published") !== "provisional") {
    // Keep published rows published; still mark rejected for the queue.
  } else {
    // Leave provisional as-is (still Custom for the user).
  }

  const { data: updatedSubmission, error: updateError } = await supabaseAdmin
    .from("fragrance_submissions")
    .update({
      status: "rejected",
      reviewed_at: new Date().toISOString(),
      reviewed_by: reviewedBy,
      promoted_fragrance_id: Number(fragrance.id),
    })
    .eq("id", submissionId)
    .select(submissionSelect)
    .single();

  if (updateError || !updatedSubmission) {
    return {
      ok: false,
      status: 500,
      message: updateError?.message ?? "Failed to reject submission.",
    };
  }

  return {
    ok: true,
    submission: mapSubmission(updatedSubmission as SubmissionRow),
    fragrance: mapFragrance(fragrance),
  };
}

export async function listSubmissionsForAdmin(
  status: "pending" | "approved" | "rejected" | "all" = "pending",
): Promise<FragranceSubmission[]> {
  let query = supabaseAdmin
    .from("fragrance_submissions")
    .select(submissionSelect)
    .order("created_at", { ascending: false })
    .limit(100);

  if (status !== "all") {
    query = query.eq("status", status);
  }

  const { data, error } = await query;
  if (error) throw new Error(error.message);
  return ((data ?? []) as SubmissionRow[]).map(mapSubmission);
}
